crates/valence_new/src/packet_stream.rs

```diff
@@ -0,0 +1,260 @@
+use std::sync::{Arc, Mutex};
+
+use bytes::BytesMut;
+use tokio::task::JoinHandle;
+use valence_protocol::packets::S2cPlayPacket;
+use valence_protocol::{DecodePacket, EncodePacket, PacketDecoder, PacketEncoder};
+
+use crate::packet::WritePacket;
+use crate::server::byte_channel::{ByteReceiver, ByteSender, TryRecvError, TrySendError};
+
+pub struct PacketStreamer {
+    pub stream: Arc<Mutex<dyn PacketStream + Send + Sync>>,
```

Comment:
Hmm this isn't quite right. We definitely don't want to be locking and unlocking a mutex every time we write or send/recv packets.

Let's see if we can untangle this a bit. First I would suggest renaming `PacketStream` to `ConnectionState` (or just `Connection`? `ClientConnection`? take your pick) since its job is to abstract away all the junk needed to maintain a real TCP connection. Here is what `Client` should look like:

```rust
struct Client {
    connection: Box<dyn ConnectionState>,
    enc: PacketEncoder,
    dec: PacketDecoder,
    ...
}
```

Notice that no matter what, we need packet encoders and decoders in `Client`. We are only using `ConnectionState` to flush all the packets in our encoder at the end of the tick, or to retrieve a hunk of bytes to put into our decoder at the beginning of the tick.

If we give `PacketEncoder` the `WritePacket` trait, we will be able to use it in most places we were previously using `PlayPacketSender`.

Here is what our real connection state should look like:

```rust
struct RealConnectionState {
    send: ByteSender,
    recv: ByteReceiver,
    permit: OwnedSemaphorePermit,
    writer_task: Option<JoinHandle<()>>,
    reader_task: JoinHandle<()>,
    handle: Handle,
}
```

There should also be a `Drop` impl on `RealConnectionState` to match what `PlayPacketSender` and `PlayPacketReceiver` were doing before.

Finally, here is what `ConnectionState` should look like:

```rust
trait ConnectionState: Send + Sync + 'static {
    fn try_send(&mut self, bytes: BytesMut) -> anyhow::Result<()>;
    fn try_recv(&mut self) -> anyhow::Result<BytesMut>;
}
```

I'm using `anyhow::Result` to hide details about the byte channel because it doesn't need to be backed by one necessarily. An error from either `try_send` or `try_recv` indicates that we should consider the client disconnected.